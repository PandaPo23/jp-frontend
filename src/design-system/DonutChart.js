import { VictoryLabel, VictoryPie, VictoryTooltip } from 'victory';
import React from 'react';

import { rem2px } from './theme';
import Card from './Card';
import theme from './theme/index';

const renderTooltip = (value, tooltipMode, t) => {
  switch (tooltipMode) {
    default:
    case 'description':
      return value.x.toUpperCase();
    case 'value':
      return Math.round(value.y);
    case 'both':
      return `${value.x.toUpperCase()} - ${Math.round(value.y)}`;
    case 'none':
      break;
  }
};

const labelShowing = (labelsMode, width, height) => {
  switch (labelsMode) {
    default:
    case 'internal':
      return Math.min(width / 2.75, height / 2.75);
    case 'external':
      break;
  }
};

const LabelAndTip = (p) => (
  <g>
    <VictoryLabel
      {...p}
      capHeight={1}
      lineHeight={theme.lineHeights.solid}
      style={{
        display: p.labelsMode,
        fill:
          p.labelsMode === 'external'
            ? theme.colors.on.background
            : p.datum.onColor,
        cursor: 'pointer',
        fontFamily: theme.fontFamily,
        fontSize: rem2px(theme.fontSizes[p.fontSize]),
      }}
    />
    <VictoryTooltip
      {...p}
      labelComponent={
        <VictoryLabel
          lineHeight={theme.lineHeights.copy}
          style={{
            fontFamily: theme.fontFamily,
            fontSize: rem2px(theme.fontSizes[p.fontSize]),
          }}
        />
      }
      style={{
        text: {
          color: p.datum.color,
        },
      }}
      text={(value) => renderTooltip(value, p.tooltipMode, p.t)}
      flyoutStyle={{
        display: p.tooltipMode,
        fill: p.datum.text,
        stroke: 'transparent',
      }}
    />
  </g>
);

LabelAndTip.defaultEvents = VictoryTooltip.defaultEvents;

const DonutChart = ({
  data,
  colors,
  tooltipMode,
  width,
  height,
  labelsMode,
  fontSize,
  angle,
  radiusFactor,
  t,
  ...props
}) => (
  <Card bg="transparent" boxShadow={0}>
    <VictoryPie
      width={width * 1.35}
      height={width * 1.35}
      data={data}
      colorScale={colors}
      innerRadius={Math.min(width / radiusFactor, height / radiusFactor)}
      labelRadius={() => labelShowing(labelsMode, width, height)}
      labels={(d) => (labelsMode === 'internal' ? d.x.toUpperCase() : d.name)}
      labelComponent={
        <LabelAndTip
          t={t}
          angle={angle}
          labelsMode={labelsMode}
          tooltipMode={tooltipMode}
          fontSize={fontSize}
          center={{ x: width / 2, y: height / 2 }}
        />
      }
      {...props}
    />
  </Card>
);

DonutChart.defaultProps = {
  width: 300,
  height: 300,
  tooltipMode: 'both',
  labelsMode: 'internal',
  fontSize: 4,
  angle: 0,
  radiusFactor: 3.85,
};

export default DonutChart;
