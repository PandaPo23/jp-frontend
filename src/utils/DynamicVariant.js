import { variant } from 'styled-system';
import PropTypes from 'prop-types';

const dynamicVariant = (variantProps) => {
  const { prop = 'variant' } = variantProps;
  const defaultVariant = variant(variantProps);

  const fn = (props) => {
    const { [prop]: variantProp } = props;

    if (!Array.isArray(variantProp)) {
      return defaultVariant;
    }
  };

  fn.propTypes = { [prop]: PropTypes.func };

  return fn;
};

export default dynamicVariant;
