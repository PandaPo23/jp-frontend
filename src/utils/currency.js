export const currencyType = {
  'en-US': 'USD',
  'de-DE': 'EUR',
  'jp-JP': 'JPY',
  'es-MX': 'MXN',
  'es-ED': 'EUR',
  'fr-FR': 'EUR',
};
