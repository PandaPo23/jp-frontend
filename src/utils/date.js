import humanizeDuration from 'humanize-duration';
import { parse as parseDuration, toSeconds } from 'iso8601-duration';
import { isBefore } from 'date-fns';

export const toValidDate = (date, lng) => {
  return new Intl.DateTimeFormat(lng).format(date);
};

export const durationToDays = (duration) =>
  parseInt(
    humanizeDuration(toSeconds(parseDuration(duration)) * 1000, {
      units: ['d'],
    }).split(' ')[0],
    10
  );

export const dfMin = (date1, date2) => (isBefore(date1, date2) ? date1 : date2);

export const dfMax = (date1, date2) => (isBefore(date1, date2) ? date2 : date1);
