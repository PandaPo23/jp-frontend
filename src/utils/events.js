export const stopEvent = (event) => {
  event.preventDefault();
  event.stopPropagation();
};

// export const processKey = ({ key, close, closeOnEscape, closeOnEnter }) => {
//   console.log({ key, close, closeOnEscape, closeOnEnter });
//   switch (key) {
//     case 'Enter':
//       if (closeOnEnter) close();
//       break;
//     case 'Escape':
//       if (closeOnEscape) close();
//       break;
//     default:
//       break;
//   }
// };
