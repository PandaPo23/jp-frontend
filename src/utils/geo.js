import polylabel from 'polylabel';

export const getDestinationsGeoJson = (destinations = []) => {
  const lines = {
    type: 'FeatureCollection',
    features: [
      {
        type: 'Feature',
        geometry: {
          type: 'LineString',
          coordinates: linearRing(
            destinations.map(({ longitude, latitude }) => [longitude, latitude])
          ),
        },
      },
    ],
  };

  const points = {
    type: 'FeatureCollection',
    features: destinations.map(({ id, name, latitude, longitude }, index) => ({
      type: 'Feature',
      geometry: {
        type: 'Point',
        coordinates: [longitude, latitude],
      },
      properties: {
        kind: 'destination',
        number: index + 1,
        id,
        name,
      },
    })),
  };

  return { lines, points };
};

export const linearRing = (coordinates = []) => {
  if (coordinates.length < 2) return coordinates;
  const [firstLon, firstLat] = coordinates[0];
  const [lastLon, lastLat] = coordinates[coordinates.length - 1];
  if (firstLon === lastLon && firstLat === lastLat) return coordinates;
  return [...coordinates, [firstLon, firstLat]];
};

export const coordinatesFromDestinations = (destinations = []) =>
  destinations.map(({ latitude, longitude }) => [longitude, latitude]);

export const centroid = (coordinates) => {
  const ring = linearRing(coordinates);
  const pos = polylabel([ring], 1.0, true);
  return pos;
};

export const REPRESENTATIVE_COORDINATE_POLICY = 'centroid';

export const representativeCoordinate = (destinations = []) => {
  const coords = coordinatesFromDestinations(destinations);
  return REPRESENTATIVE_COORDINATE_POLICY === 'first'
    ? coords[0]
    : centroid(coords);
};

export const getBoundsFromDestinations = (destinations = []) => {
  const sw = [181, 91];
  const ne = sw.map((x) => -x);
  return destinations.reduce(
    ([sw, ne], { longitude, latitude }) => [
      [Math.min(sw[0], longitude), Math.min(sw[1], latitude)],
      [Math.max(ne[0], longitude), Math.max(ne[1], latitude)],
    ],
    [sw, ne]
  );
};

export const getBoundsFromTrips = (trips = []) => {
  return getBoundsFromDestinations(
    trips.reduce((z, t) => z.concat(t.destinations), [])
  );
};

export const getGeoJSONForTrips = (trips = []) => {
  const gj = {
    points: {
      type: 'FeatureCollection',
      features: trips.reduce((acc, { id, destinations, name }, index) => {
        acc.push({
          type: 'Feature',
          geometry: {
            type: 'Point',
            coordinates: representativeCoordinate(destinations),
          },
          properties: {
            kind: 'trip',
            number: index + 1,
            id,
            name,
          },
        });
        return acc;
      }, []),
    },
  };
  return gj;
};

export const getGeoJSONForDestination = (destination) => {
  const point = {
    type: 'Feature',
    geometry: {
      type: 'Point',
      coordinates: [destination.longitude, destination.latitude],
    },
    properties: {
      kind: 'destination',
      id: destination.id,
      name: destination.name,
    },
  };

  return { point };
};

export const getBoundsOfDestinationFromTrips = (destination, trip) => {
  const { longitude, latitude } = destination;
  const radius = trip.destinations.reduce(
    (radius, dest) =>
      dest.id === destination.id
        ? radius
        : Math.min(
            radius,
            Math.abs(dest.longitude - longitude),
            Math.abs(dest.latitude - latitude)
          ),
    180
  );
  return [
    [longitude - radius, latitude - radius],
    [longitude + radius, latitude + radius],
  ];
};

export const getBoundsOfCountry = (country) => {
  const bb = country.boundingBox;
  return {
    country: {
      type: 'Feature',
      geometry: {
        type: 'Point',
        coordinates: [[bb[0], bb[1]], [bb[2], bb[3]]],
      },
    },
  };
};
