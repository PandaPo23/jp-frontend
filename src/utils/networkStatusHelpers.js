import { toJS } from 'mobx';
import fp from 'lodash/fp';

export const isLoadingTrip = fp.compose(
  Boolean,
  fp.find({ type: 'tripDetails' }),
  toJS
);

export const isLoadingTrips = fp.compose(
  Boolean,
  fp.find({ type: 'matchedTrips' }),
  toJS
);
