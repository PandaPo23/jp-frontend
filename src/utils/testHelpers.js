import React from 'react';
import { mount, shallow } from 'enzyme';
import { ThemeProvider } from 'emotion-theming';
import theme from '../design-system/theme';

export const mountWithTheme = (component) =>
  mount(<ThemeProvider theme={theme}>{component}</ThemeProvider>);

export const shallowWithTheme = (component) =>
  shallow(<ThemeProvider theme={theme}>{component}</ThemeProvider>);
