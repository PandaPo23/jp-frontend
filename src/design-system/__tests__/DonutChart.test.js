import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import { shallowWithTheme } from '../../utils/testHelpers';
import DonutChart from '../DonutChart';

let colors = ['#69a6f1', '#64e5aa', '#cea0f8', '#f1df86'];
let props;
let component;
let travelStyles;
describe('DonutChart component', () => {
  beforeEach(() => {
    travelStyles = [
      {
        color: '#69a6f1',
        label: 'OC',
        name: 'Oceanist',
        onColor: '#ffffff',
        x: 'oc',
        y: 10.050251256281408,
      },
      {
        color: '#64e5aa',
        label: 'AN',
        name: 'Active Naturist',
        onColor: '#ffffff',
        x: 'an',
        y: 17.08542713567839,
      },
      {
        color: '#a3f490',
        label: 'CN',
        name: 'Chilled Naturist',
        onColor: '#ffffff',
        x: 'cn',
        y: 46.733668341708544,
      },
      {
        color: '#cea0f8',
        label: 'CU',
        name: 'Culturist',
        onColor: '#ffffff',
        x: 'cu',
        y: 22.613065326633166,
      },
      {
        color: '#f1df86',
        label: 'PP',
        name: 'Party Purist',
        onColor: '#ffffff',
        x: 'pp',
        y: 3.5175879396984926,
      },
    ];
  });
  it('should always render a DonutChart', () => {
    const localProps = {
      data: travelStyles,
      colors,
      view: 'pie',
    };
    const wrapper = shallowWithTheme(<DonutChart {...localProps} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
