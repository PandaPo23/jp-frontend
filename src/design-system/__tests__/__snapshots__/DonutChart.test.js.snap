// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DonutChart component should always render a DonutChart 1`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;
